grimoire/core/styling/palette.py

```diff
@@ -0,0 +1,148 @@
+
+from gdpc import Block
+
+from grimoire.core.assets.asset import Asset
+from grimoire.core.styling.blockform import BlockForm
+from grimoire.core.styling.colors import MinecraftColor
+from grimoire.core.styling.materials.dithering import DitheringPattern
+from grimoire.core.styling.materials.material import Material, MaterialParameters
+
+
+class MaterialRole(Enum):
+    primary_wall = "primary_wall"
+    secondary_wall = "secondary_wall"
+
+    primary_roof = "primary_roof"
+    secondary_roof = "secondary_roof"
+
+    primary_stone = "primary_stone"
+    secondary_stone = "secondary_stone"
+
+    primary_wood = "primary_wood"
+    secondary_wood = "secondary_wood"
+
+    pillar = "pillar"
+
+
+class _ResolutionPriority:
+    order: list[MaterialRole]
+    next_in_line: dict[MaterialRole, MaterialRole]
+
+    def __init__(
+        self, order: list[MaterialRole], next_in_line: dict[MaterialRole, MaterialRole]
+    ):
+        self.order = order
+        self.next_in_line = next_in_line
+
+    def get_next_in_line(self, role: MaterialRole) -> MaterialRole | None:
+        if role in self.next_in_line:
+            return self.next_in_line[role]
+
+        return None
+
+
+class ResolutionPriority(Enum):
+    stone = _ResolutionPriority(
+        order=[
+            MaterialRole.pillar,
+            MaterialRole.primary_wall,
+            MaterialRole.secondary_wall,
+            MaterialRole.primary_roof,
+            MaterialRole.secondary_roof,
+            MaterialRole.primary_stone,
+            MaterialRole.secondary_stone,
+            MaterialRole.primary_wood,
+            MaterialRole.secondary_wood,
+        ],
+        next_in_line={
+            MaterialRole.primary_wall: MaterialRole.primary_stone,
+            MaterialRole.secondary_wall: MaterialRole.secondary_stone,
+            MaterialRole.primary_roof: MaterialRole.primary_stone,
+            MaterialRole.secondary_roof: MaterialRole.secondary_stone,
+            MaterialRole.primary_stone: MaterialRole.primary_wood,
+            MaterialRole.secondary_stone: MaterialRole.secondary_wood,
+            MaterialRole.primary_wood: MaterialRole.primary_stone,
+            MaterialRole.secondary_wood: MaterialRole.secondary_stone,
+            MaterialRole.pillar: MaterialRole.primary_stone,
+        },
+    )
+
+    def order(self) -> list[MaterialRole]:
+        return self.value.order
+
+    def get_next_in_line(self, role: MaterialRole) -> MaterialRole | None:
+        return self.value.get_next_in_line(role)
+
+
+class Palette(Asset):
+    resolution_priority: ResolutionPriority
+    materials: dict[MaterialRole, Material]
+
+    # Colors
+    primary_color: MinecraftColor
+    secondary_color: MinecraftColor
+
+    def find_role(self, block: Block) -> MaterialRole | None:
+        for role in self.resolution_priority.order():
+            if role in self.materials and self.materials[role].has_block(block):
+                return role
+
+        return None
```

Comment:
```py
return next(
            (
                role
                for role in self.resolution_priority.order()
                if role in self.materials and self.materials[role].has_block(block)
            ),
            None,
        )
```

Reply:
I feel like this is less clear than the original code

Reply:
But you are just returning one element at a time, ending with "None", yes?